File: 0329.LongestIncreasingPathinaMatrix.py
class Solution(object):
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        def dfs(i, j):
            for x, y in ((i-1, j), (i+1, j), (i, j-1), (i, j+1)):
                if 0 <= x < m and 0 <= y < n and matrix[x][y] > matrix[i][j]:
                    if not step[x][y]:
                        dfs(x, y)
                    step[i][j] = max(step[i][j], step[x][y]+1)
            if not step[i][j]:
                step[i][j] = 1

        if not matrix:
            return 0
        m, n = len(matrix), len(matrix[0])
        step = [[0]*n for i in xrange(m)]
        for i in xrange(len(matrix)):
            for j in xrange(len(matrix[0])):
                if not step[i][j]:
                    dfs(i, j)
        return max([max(t) for t in step])

File: 0329.LongestIncreasingPathinaMatrix.go
func longestIncreasingPath(matrix [][]int) int {
    if len(matrix) == 0 || len(matrix[0]) == 0 {
        return 0
    }
    ans := 0
    m, n := len(matrix), len(matrix[0])
    depth := make([][]int, m)
    for i := range depth {
        depth[i] = make([]int, n)
    }
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            ans = max(ans, dfs(matrix, depth, i, j))
        }
    }
    return ans
}

func dfs(matrix, depth [][]int, i, j int) int {
    m, n := len(matrix), len(matrix[0])
    if depth[i][j] != 0 {
        return depth[i][j]
    }
    dir := [][2]int{[2]int{-1, 0}, [2]int{0, -1}, [2]int{0, 1}, [2]int{1, 0}}
    for _, d := range dir {
        x, y := i+d[0], j+d[1]
        if x >= 0 && x < m && y >= 0 && y < n && matrix[x][y] > matrix[i][j] {
            depth[i][j] = max(depth[i][j], dfs(matrix, depth, x, y))
        }
    }
    depth[i][j]++
    return depth[i][j]
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}